### PR TITLE
Support ONNX runtime in RunInference API 

### PR DESCRIPTION
**Pull request by ziqi-ma**
**Please** add a meaningful description for your change here
Addresses #22972 

Adding onnx support to beam. ONNX Runtime is a popular framework to accelerate ML inference:
https://onnxruntime.ai/. It supports a broad range of model frameworks including PyTorch, Tensorflow/Keras, TFLite, scikit-learn, etc.. This PR lets you call an onnx model from beam by passing in model path.

Specifically:
- Added onnx_inference.py which is the onnx API.
- Added onnx_inference_test.py which contains unit tests for onnx models based on PyTorch, tensor flow, and sklearn.
- Added onnx_inference_it_test.py which contains integration test for sentiment classification with onnx version of RoBERTa.
- Added examples/onnx_sentiment_classification.py which is an example of using an onnx version of RoBERTa for sentiment classification, also used in the integration test.
- Configured test with python 3.8 in tox and gradle.

Thank you for your contribution! Follow this checklist to help us incorporate your contribution quickly and easily:

 - [ ] Mention the appropriate issue in your description (for example: `addresses #123`), if applicable. This will automatically add a link to the pull request in the issue. If you would like the issue to automatically close on merging the pull request, comment `fixes #<ISSUE NUMBER>` instead.
 - [ ] Update `CHANGES.md` with noteworthy changes.
 - [ ] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

See the [Contributor Guide](https://beam.apache.org/contribute) for more tips on [how to make review process smoother](https://beam.apache.org/contribute/get-started-contributing/#make-the-reviewers-job-easier).

To check the build health, please visit [https://github.com/apache/beam/blob/master/.test-infra/BUILD_STATUS.md](https://github.com/apache/beam/blob/master/.test-infra/BUILD_STATUS.md)

GitHub Actions Tests Status (on master branch)
------------------------------------------------------------------------------------------------
[Build python source distribution and wheels](https://github.com/apache/beam/actions?query=workflow%3A%22Build+python+source+distribution+and+wheels%22+branch%3Amaster+event%3Aschedule)
[Python tests](https://github.com/apache/beam/actions?query=workflow%3A%22Python+Tests%22+branch%3Amaster+event%3Aschedule)
[Java tests](https://github.com/apache/beam/actions?query=workflow%3A%22Java+Tests%22+branch%3Amaster+event%3Aschedule)
[Go tests](https://github.com/apache/beam/actions?query=workflow%3A%22Go+tests%22+branch%3Amaster+event%3Aschedule)

See [CI.md](https://github.com/apache/beam/blob/master/CI.md) for more information about GitHub Actions CI.

### CODE REVIEW

**Comment by AnandInguva** [comment]
One question: Do we need GPUs to run the ONNX unit tests? I have very little about ONNX but I will have a read about it during the review.

**Comment by ziqi-ma** [comment]
> One question: Do we need GPUs to run the ONNX unit tests? I have very little about ONNX but I will have a read about it during the review.

In the code I did specify ONNX provider in a way (as a priority list) such that if run on NVIDIA GPUs they should run the gpu version but on cpu it should run the cpu version. But I ran my tests on CPU - if we want to make sure this works properly in a GPU environment (and which type of GPU also matters) then I need to test in those environments too.

File: sdks/python/apache_beam/ml/inference/onnx_inference.py
Comment (AnandInguva) [inline sdks/python/apache_beam/ml/inference/onnx_inference.py:84]:
```suggestion
class OnnxModelHandlerNumpy(ModelHandler[numpy.ndarray,
```

File: sdks/python/apache_beam/ml/inference/onnx_inference.py
Comment (AnandInguva) [inline sdks/python/apache_beam/ml/inference/onnx_inference.py:93]: Can we also add a note saying the inputs to ONNXModelHandler should be of the same sizes. If different size inputs are expected, user have to explicitly declare batch size as 1.

File: sdks/python/apache_beam/ml/inference/onnx_inference.py
Comment (AnandInguva) [inline sdks/python/apache_beam/ml/inference/onnx_inference.py:137]: Can we follow this for the num_bytes ?https://github.com/apache/beam/blob/95e53916b6c9de6052ce8fc8409a53bad3d9d517/sdks/python/apache_beam/ml/inference/tensorrt_inference.py#L307

File: sdks/python/apache_beam/ml/inference/onnx_inference.py
Comment (AnandInguva) [inline sdks/python/apache_beam/ml/inference/onnx_inference.py:36]: I don't think we use this anywhere. Do you want to mark this experimental? If yes, why?

File: sdks/python/apache_beam/ml/inference/onnx_inference.py
Comment (AnandInguva) [inline sdks/python/apache_beam/ml/inference/onnx_inference.py:38]: Can we remove unused dependencies imports?

File: sdks/python/apache_beam/ml/inference/onnx_inference.py
Comment (AnandInguva) [inline sdks/python/apache_beam/ml/inference/onnx_inference.py:101]:
```suggestion
      inference_fn: The inference function to use.
```
```suggestion
      inference_fn: The inference function to use on RunInference calls..
```

File: sdks/python/apache_beam/ml/inference/onnx_inference_test.py
Comment (AnandInguva) [inline sdks/python/apache_beam/ml/inference/onnx_inference_test.py:176]: uses_pytorch markers runs Pytorch tests, which requires [https://github.com/apache/beam/blob/95e53916b6c9de6052ce8fc8409a53bad3d9d517/sdks/python/apache_beam/ml/inference/torch_tests_requirements.txt](requirements) file to be installed for dependencies. We either add onnx dependencies to that requirements file.

Same goes with tensorflow. https://github.com/apache/beam/blob/95e53916b6c9de6052ce8fc8409a53bad3d9d517/sdks/python/apache_beam/examples/inference/tfx_bsl/requirements.txt

File: start-build-env-onnx.sh
Comment (AnandInguva) [inline start-build-env-onnx.sh:1]: For my knowledge, why do we need this file?

**Comment by ziqi-ma** [comment]
> Hi, once the PR is ready for another review, comment on this with `PTAL @<username>`.
> Thanks for the starting this.

PTAL @AnandInguva

**Comment by AnandInguva** [comment]
> > Hi, once the PR is ready for another review, comment on this with `PTAL @<username>`.
> > Thanks for the starting this.
> 
> PTAL @AnandInguva [I'm still working on the gradle onnxtest (if that is the right direction to go), but in the meantime could you let me know how to put data at gcp location for integration test? I think maybe that should come before the gradle part.]

Hi, `gs://apache-beam-samples/run_inference` should be open to public. You can upload your required files/folder here for testing.

File: sdks/python/apache_beam/ml/inference/onnx_inference_test.py
Comment (AnandInguva) [inline sdks/python/apache_beam/ml/inference/onnx_inference_test.py:36]: For the tests to run, we can have a tox command like https://github.com/apache/beam/blob/9f7c55f40d3d5bba8b5d1a428f6ff42ef8082bbf/sdks/python/tox.ini#L317 which could install all the dependencies required to run this file.

File: sdks/python/apache_beam/ml/inference/onnx_inference.py
Comment (AnandInguva) [inline sdks/python/apache_beam/ml/inference/onnx_inference.py:73]: I am not sure if passing inference_args like this is the right way. 

Let us assume model predict call accepts 
```
def predict(inputs, dropout=True):
   .... 
   # some processing on inputs, calculate output_1, output_2
   if dropout:
         return output_1
   else:
         return output_2
```
We pass the dropout argument via inference_args: `inference_args = {'dropout`: True}. How would we pass this extra input to the `inference_session` ?

**Comment by AnandInguva** [comment]
Hi, I will take another review this weekend.  If you have some time, you can fix the formatting issues. 

image

**Comment by ziqi-ma** [comment]
> Hi, I will take another review this weekend. If you have some time, you can fix the formatting issues.
> 
> image

Fixed yaps

**Comment by ziqi-ma** [comment]
> > Hi, I will take another review this weekend. If you have some time, you can fix the formatting issues.
> > image
> 
> Fixed yaps

**Comment by ziqi-ma** [comment]
> > > Hi, once the PR is ready for another review, comment on this with `PTAL @<username>`.
> > > Thanks for the starting this.
> > 
> > 
> > PTAL @AnandInguva [I'm still working on the gradle onnxtest (if that is the right direction to go), but in the meantime could you let me know how to put data at gcp location for integration test? I think maybe that should come before the gradle part.]
> 
> Hi, `gs://apache-beam-samples/run_inference` of project `apache-beam-testing` should be open to public. You can upload your required files/folder here for testing.

Yes I am able to access apache-beam-samples/run_inference. However, it seems like the other tests use gs://apache-beam-ml/models/ or gs://apache-beam-ml/datasets, but I'm getting access denied. Is there a way to get access? Or should I just put the test data/models in examples?

**Comment by AnandInguva** [comment]
> > > > Hi, once the PR is ready for another review, comment on this with `PTAL @<username>`.
> > > > Thanks for the starting this.
> > > 
> > > 
> > > PTAL @AnandInguva [I'm still working on the gradle onnxtest (if that is the right direction to go), but in the meantime could you let me know how to put data at gcp location for integration test? I think maybe that should come before the gradle part.]
> > 
> > 
> > Hi, `gs://apache-beam-samples/run_inference` of project `apache-beam-testing` should be open to public. You can upload your required files/folder here for testing.
> 
> For apache-beam-samples/run_inference, I am able to read but not write, getting the error below: User [[ziqima20@gmail.com](mailto:ziqima20@gmail.com)] does not have permission to access b instance [apache-beam-samples] (or it may not exist): [ziqima20@gmail.com](mailto:ziqima20@gmail.com) does not have storage.objects.create access to the Google Cloud Storage object. Permission 'storage.objects.create' denied on resource (or it may not exist).
> 
> However, it seems like the other tests use gs://apache-beam-ml/models/ and gs://apache-beam-ml/datasets. For these I do not have read access.

Yes that is true. You can share the files with me and I can upload them to the GCS bucket for you. You can share them to me at anandinguva98@gmail.com

File: sdks/python/tox.ini
Comment (AnandInguva) [inline sdks/python/tox.ini:330]: You would need to add this to `sdks/python/test-suites/tox/py38/build.gradle` as well.

File: sdks/python/tox.ini
Comment (AnandInguva) [inline sdks/python/tox.ini:332]: Can we just pin `onnxruntime==1.13.1`?

File: sdks/python/tox.ini
Comment (AnandInguva) [inline sdks/python/tox.ini:330]:
```suggestion
[testenv:py{37,38,39,310}-onnx-{113}]
```

**Comment by AnandInguva** [comment]
cc: @damccorm @riteshghorse @jrmccluskey for reviews.

**Comment by jrmccluskey** [comment]
Run PythonDocs PreCommit

**Comment by jrmccluskey** [comment]
Information on running formatting and linting locally is here: https://cwiki.apache.org/confluence/display/BEAM/Python+Tips#PythonTips-LintandFormattingChecks

File: sdks/python/tox.ini
Comment (jrmccluskey) [reply sdks/python/tox.ini:332]: +1

**Comment by ziqi-ma** [comment]
> > > > > Hi, once the PR is ready for another review, comment on this with `PTAL @<username>`.
> > > > > Thanks for the starting this.
> > > > 
> > > > 
> > > > PTAL @AnandInguva [I'm still working on the gradle onnxtest (if that is the right direction to go), but in the meantime could you let me know how to put data at gcp location for integration test? I think maybe that should come before the gradle part.]
> > > 
> > > 
> > > Hi, `gs://apache-beam-samples/run_inference` of project `apache-beam-testing` should be open to public. You can upload your required files/folder here for testing.
> > 
> > 
> > For apache-beam-samples/run_inference, I am able to read but not write, getting the error below: User [[ziqima20@gmail.com](mailto:ziqima20@gmail.com)] does not have permission to access b instance [apache-beam-samples] (or it may not exist): [ziqima20@gmail.com](mailto:ziqima20@gmail.com) does not have storage.objects.create access to the Google Cloud Storage object. Permission 'storage.objects.create' denied on resource (or it may not exist).
> > However, it seems like the other tests use gs://apache-beam-ml/models/ and gs://apache-beam-ml/datasets. For these I do not have read access.
> 
> Yes that is true. You can share the files with me and I can upload them to the GCS bucket for you. You can share them to me at [anandinguva98@gmail.com](mailto:anandinguva98@gmail.com)

Hi @AnandInguva - This PR is ready on my end - I added all the gap-related tests using my own gcp storage. Could you help copy the files below to the beam gcp buckets? (My bucket is open to public)

gs://ziqi-bucket1/torch_2xplus5_onnx   -> gs://apache-beam-ml/models/torch_2xplus5_onnx
gs://ziqi-bucket1/tf_2xplus5_onnx          -> gs://apache-beam-ml/models/tf_2xplus5_onnx
gs://ziqi-bucket1/skl_2xplus5_onnx        -> gs://apache-beam-ml/models/skl_2xplus5_onnx

gs://ziqi-bucket1/sentiment_classification_input.txt   -> gs://apache-beam-ml/testing/inputs/onnx/sentiment_classification_input.txt
gs://ziqi-bucket1/roberta_sentiment_classification.onnx  -> gs://apache-beam-ml/models/roberta_sentiment_classification.onnx
gs://ziqi-bucket1/sentiment_classification_expected_output.txt -> gs://apache-beam-ml/testing/expected_outputs/test_onnx_run_inference_roberta_sentiment_classification_actuals.txt

Thanks!

**Comment by AnandInguva** [comment]
> 

Hi, I uploaded them.

File: sdks/python/apache_beam/examples/inference/onnx_sentiment_classification.py
Comment (AnandInguva) [inline sdks/python/apache_beam/examples/inference/onnx_sentiment_classification.py:126]: Can we make sure there `input` is required arg and remove this part of code? Reduces complexity.

File: sdks/python/apache_beam/examples/inference/onnx_sentiment_classification.py
Comment (AnandInguva) [inline sdks/python/apache_beam/examples/inference/onnx_sentiment_classification.py:106]:
```suggestion
    """Wrapper to OnnxModelHandlerNumpy to limit batch size to 1.
```

File: sdks/python/apache_beam/examples/inference/onnx_sentiment_classification.py
Comment (AnandInguva) [inline sdks/python/apache_beam/examples/inference/onnx_sentiment_classification.py:151]:
```suggestion
  _  = output | "WriteOutput" >> beam.io.WriteToText(
```

Comment (AnandInguva) [reply sdks/python/apache_beam/examples/inference/onnx_sentiment_classification.py:151]: Can we try to decrease/solve pylint errors instead of ignoring them?

**Review by AnandInguva — COMMENTED** [review]
Most of this looks good to me. 

I would clean up the lint, formatting errors for the checks to go green and we can start finalizing the PR.

File: sdks/python/apache_beam/ml/inference/onnx_inference_test.py
Comment (AnandInguva) [inline sdks/python/apache_beam/ml/inference/onnx_inference_test.py:441]: Can we wrap the long line using parentheses instead of back slash? 

```
For reference: From PEP 8, style guild https://peps.python.org/pep-0008/#maximum-line-length

The preferred way of wrapping long lines is by using Python’s implied line continuation inside parentheses, brackets and braces. Long lines can be broken over multiple lines by wrapping expressions in parentheses. These should be used in preference to using a backslash for line continuation.
```

**Comment by ziqi-ma** [comment]
> Most of this looks good to me.
> 
> I would clean up the lint, formatting errors for the checks to go green and we can start finalizing the PR.

Hi -thanks. I fixed the formatting errors but seems like the 2 remaining failures are about files/tests that should not be touched by this PR?

- codecov/patch is complaining 0.00% diff hit (and it seems to be one line not covered in sdks/python/apache_beam/runners/worker/bundle_processor.py? (I'm not sure how to cover it?)
- py310 test is failing due to a segfault, but I only added onnx tests for py38?

Would be great if you could give some pointers regarding these, thanks!

**Comment by AnandInguva** [comment]
you can ignore codecov and the other error is not relevant to this PR and in general...could be flaky.

File: sdks/python/apache_beam/ml/inference/onnx_inference.py
Comment (AnandInguva) [inline sdks/python/apache_beam/ml/inference/onnx_inference.py:41]: #25200 moved this method to apache_beam/ml/inference/utils.py. Can you refactor the change to it?

**Review by AnandInguva — APPROVED** [review]

**Comment by AnandInguva** [comment]
@ziqi-ma Hi, is it ready for the final review? 

I am not a committer so passing this @jrmccluskey @damccorm @riteshghorse @tvalentyn 

Thanks again for the changes.

File: sdks/python/tox.ini
Comment (riteshghorse) [inline sdks/python/tox.ini:348]:
```suggestion
  # Run all ONNX unit tests
```

**Review by riteshghorse — APPROVED** [review]
one minor nit, LGTM

**Comment by riteshghorse** [comment]
Since you've already added an example, please add a section to [README](https://github.com/apache/beam/blob/master/sdks/python/apache_beam/examples/inference/README.md) on how to run it. You can add it in a new PR as well if you'd like.

**Review by damccorm — APPROVED** [review]

**Comment by damccorm** [comment]
LGTM - we can merge once Ritesh's comments are responded to

**Comment by ziqi-ma** [comment]
> Since you've already added an example, please add a section to [README](https://github.com/apache/beam/blob/master/sdks/python/apache_beam/examples/inference/README.md) on how to run it. You can add it in a new PR as well if you'd like.

Added

**Comment by ziqi-ma** [comment]
> @ziqi-ma Hi, is it ready for the final review?
> 
> I confirm the tests are running for onnx. link: https://ci-beam.apache.org/job/beam_PreCommit_Python_Coverage_Commit/398/testReport/apache_beam.ml.inference.onnx_inference_test/ I am not a committer so passing this @jrmccluskey @damccorm @riteshghorse @tvalentyn
> 
> Thanks again for the changes.
> 
> Also, can you add this feature to the CHANGES.md defined at https://github.com/apache/beam/blob/master/CHANGES.md?
> 
> Something like
> 
> https://github.com/apache/beam/blob/64e40d2c018f8e906f4bec32ef67f02734a95721/CHANGES.md?plain=1#L57
> 
> I would like to get this before Feb 22nd so that it can be included in next release 2.46.0

Hi - I have fixed all comments. Think this is ready to merge.

**Review by damccorm — APPROVED** [review]
This is awesome, thank you!

**Comment by damccorm** [comment]
retest this please

**Comment by damccorm** [comment]
(above comment was for the CI bot) - I'm going to let the precommit checks run to completion to make sure they are passing, then I will merge

**Comment by damccorm** [comment]
Thanks @ziqi-ma!

**Comment by Abacn** [comment]
It breaks Python PostCommits `hdfsIntegrationTest` task, error message is (taken from https://ci-beam.apache.org/view/PostCommit/job/beam_PostCommit_Python310/470/consoleFull)

```
04:11:40 test_1      | Traceback (most recent call last):
04:11:40 test_1      |   File "/usr/local/bin/tox", line 8, in <module>
04:11:40 test_1      |     sys.exit(cmdline())
04:11:40 test_1      |   File "/usr/local/lib/python3.10/site-packages/tox/session/__init__.py", line 42, in cmdline
04:11:40 test_1      |     main(args)
04:11:40 test_1      |   File "/usr/local/lib/python3.10/site-packages/tox/session/__init__.py", line 62, in main
04:11:40 test_1      |     config = load_config(args)
04:11:40 test_1      |   File "/usr/local/lib/python3.10/site-packages/tox/session/__init__.py", line 78, in load_config
04:11:40 test_1      |     config = parseconfig(args)
04:11:40 test_1      |   File "/usr/local/lib/python3.10/site-packages/tox/config/__init__.py", line 262, in parseconfig
04:11:40 test_1      |     ParseIni(config, config_file, content)
04:11:40 test_1      |   File "/usr/local/lib/python3.10/site-packages/tox/config/__init__.py", line 1126, in __init__
04:11:40 test_1      |     raise tox.exception.ConfigError(
04:11:40 test_1      | tox.exception.ConfigError: ConfigError: py{38}-onnx-{113} failed with ConfigError: substitution key '38' not found at Traceback (most recent call last):
04:11:40 test_1      |   File "/usr/local/lib/python3.10/site-packages/tox/config/__init__.py", line 1102, in run
04:11:40 test_1      |     results[name] = cur_self.make_envconfig(name, section, subs, config)
04:11:40 test_1      |   File "/usr/local/lib/python3.10/site-packages/tox/config/__init__.py", line 1226, in make_envconfig
04:11:40 test_1      |     res = meth(env_attr.name, env_attr.default, replace=replace)
04:11:40 test_1      |   File "/usr/local/lib/python3.10/site-packages/tox/config/__init__.py", line 1407, in getpath
04:11:40 test_1      |     path = self.getstring(name, defaultpath, replace=replace)
04:11:40 test_1      |   File "/usr/local/lib/python3.10/site-packages/tox/config/__init__.py", line 1486, in getstring
04:11:40 test_1      |     x = self._replace_if_needed(x, name, replace, crossonly)
04:11:40 test_1      |   File "/usr/local/lib/python3.10/site-packages/tox/config/__init__.py", line 1491, in _replace_if_needed
04:11:40 test_1      |     x = self._replace(x, name=name, crossonly=crossonly)
04:11:40 test_1      |   File "/usr/local/lib/python3.10/site-packages/tox/config/__init__.py", line 1517, in _replace
04:11:40 test_1      |     replaced = Replacer(self, crossonly=crossonly).do_replace(value)
04:11:40 test_1      |   File "/usr/local/lib/python3.10/site-packages/tox/config/__init__.py", line 1557, in do_replace
04:11:40 test_1      |     expanded = substitute_once(value)
04:11:40 test_1      |   File "/usr/local/lib/python3.10/site-packages/tox/config/__init__.py", line 1551, in substitute_once
04:11:40 test_1      |     return self.RE_ITEM_REF.sub(self._replace_match, x)
04:11:40 test_1      |   File "/usr/local/lib/python3.10/site-packages/tox/config/__init__.py", line 1598, in _replace_match
04:11:40 test_1      |     return self._replace_substitution(match)
04:11:40 test_1      |   File "/usr/local/lib/python3.10/site-packages/tox/config/__init__.py", line 1633, in _replace_substitution
04:11:40 test_1      |     val = self._substitute_from_other_section(sub_key)
04:11:40 test_1      |   File "/usr/local/lib/python3.10/site-packages/tox/config/__init__.py", line 1627, in _substitute_from_other_section
04:11:40 test_1      |     raise tox.exception.ConfigError("substitution key {!r} not found".format(key))
04:11:40 test_1      | tox.exception.ConfigError: ConfigError: substitution key '38' not found
04:11:40 test_1      | 
```
see #25443 for details

**Pull request merged**
